python/ray/workflow/step_executor.py

```diff
@@ -225,6 +216,47 @@ def execute_workflow(workflow: "Workflow") -> "WorkflowExecutionResult":
     return result
 
 
+@dataclass
+class InplaceReturnedWorkflow:
+    """Hold information about a workflow returned from an inplace step."""
+
+    # The returned workflow.
+    workflow: Workflow
+    # The dict that contains the context of the inplace returned workflow.
+    context: Dict
+
+
+def execute_workflow(workflow: Workflow) -> "WorkflowExecutionResult":
+    """Execute workflow.
+
+    This function also performs tail-recursion optimization for inplace
+    workflow steps.
+
+    Args:
+        workflow: The workflow to be executed.
+    Returns:
+        An object ref that represent the result.
+    """
+    # Tail recursion optimization.
+    context = {}
+    while True:
+        with workflow_context.fork_workflow_step_context(**context):
+            result = _execute_workflow(workflow)
+        if not isinstance(result.persisted_output, InplaceReturnedWorkflow):
+            break
+        workflow = result.persisted_output.workflow
+        context = result.persisted_output.context
+        # prevent leaking of workflow refcount
+        del result
```

Comment:
Could you add more detailed comment here about why do we need this?

Reply:
Thanks for pointing out it! I think we no longer need this line now.